### DISCUSSION TITLE
Avoiding jagged edges with `mapim`?

### DISCUSSION THREAD

**Discussion by GavinJoyce** [libvips/ruby-vips#323] [Q&A] [answered]
Hi,

I'm following [this example](https://stackoverflow.com/a/50219715) to achieve perspective distortion with `mapim`.

It's working great, except for one small issue with jagged edges which I haven't yet figured out how to avoid. Perhaps someone knows how to avoid the jagged edges?

**output.png:**
![output](https://user-images.githubusercontent.com/2526/147763303-545fdd07-7057-4f1d-9493-1b815df5fea6.png)

**input.png:**
![input](https://user-images.githubusercontent.com/2526/147763385-a158f5a2-6f2a-4058-af64-40c890572a43.png)

```rb
require "vips"

image = Vips::Image.new_from_file "input.png"

t = [1.73945282, -0.41595611, -116.35426798, 5.97741712e-01, 2.47280054e+00, -6.75608244e+02, -1.59808198e-06, 4.97586518e-06]
i = Vips::Image.xyz image.width, image.height
x = (i[0] * t[0] + i[1] * t[1] + t[2]) / (i[0] * t[6] + i[1] * t[7] + 1)
y = (i[0] * t[3] + i[1] * t[4] + t[5]) / (i[0] * t[6] + i[1] * t[7] + 1)

m = x.bandjoin y
distorted_image = image.mapim m

distorted_image.write_to_file "output.png"
```

**Comment by GavinJoyce** [comment]
Seems similar to https://github.com/libvips/php-vips/issues/58 ("affine should have a background parameter, and not have jaggy edges")

**Comment by jcupitt** [comment, marked as answer]
Hi @GavinJoyce,

You could put a one-pixel black border around the source. Eg.:

```ruby
#!/usr/bin/ruby

require "vips"

image = Vips::Image.new_from_file ARGV[0]
image = image.embed 1, 1, image.width + 2, image.height + 2

t = [1.73945282, -0.41595611, -116.35426798, 5.97741712e-01, 2.47280054e+00, -6.75608244e+02, -1.59808198e-06, 4.97586518e-06]
i = Vips::Image.xyz image.width, image.height
x = (i[0] * t[0] + i[1] * t[1] + t[2]) / (i[0] * t[6] + i[1] * t[7] + 1)
y = (i[0] * t[3] + i[1] * t[4] + t[5]) / (i[0] * t[6] + i[1] * t[7] + 1)

m = x.bandjoin y
distorted_image = image.mapim m

distorted_image.write_to_file ARGV[1]
```

To make:

![x](https://user-images.githubusercontent.com/580843/147764786-7cda8ca5-f708-4de9-b2db-8665be3c9327.png)

**Comment by jcupitt** [comment]
... but you're right, improving `mapim`'s edge handling would be better.

**Comment by GavinJoyce** [comment]
Thanks for the workaround @jcupitt, that works a treat

**Reply by jcupitt** [reply]
I made an enhancement issue to fix this properly. Thanks for the suggestion.

**Comment by jcupitt** [comment]
... this is fixed in master and will be in 8.13.
